lib/hanami/action/contract.rb

```diff
@@ -0,0 +1,134 @@
+# frozen_string_literal: true
+
+module Hanami
+  class Action
+    # A wrapper for defining validation rules using Dry::Validation. This class essentially
+    # wraps a Dry::Validation::Contract and acts as a proxy to actually use Dry gem
+    #
+    # Defined via the `contract` block in an action class.
+    # Although more complex domain-specific validations, or validations concerned with things such as uniqueness
+    # are usually better performed at layers deeper than your HTTP actions, Contract still provides helpful features
+    # that you can use without contravening the advice form above.
+    #
+    # @since 2.2.0
+    class Contract
```

Comment:
So if I trace along the new logic in the PR:

In an action where a user defines a contract with `contract do`, then an instance of `Hanami::Action::Contract` will become  becomes the `params` object that we provide to the `handle` method as part of the `request`.

Am I understanding this correctly?

If so, since this `params` object can now be an instance of either `Hanami::Action::Contract` _or_ `Hanami::Action::Params`, I think we want to make sure they have the same interface.

Why is this? I think it would be perfectly acceptable for a user within a single app to have `params do` in some action classes, and `contract do` in others. So this means that across a single app, the user will get different kinds of `params` instances, but that user should expect those objects to behave identically.

Right now `Hanami::Action::Params` implements a few things that the new `Hanami::Action::Contract` does not:

- `#raw`
- `#error_messages`
- `#deconstruct_keys`
- `#get`, `#dig` (via `BaseParams`)
- `#each` (via `BaseParams`)

We need to make these consistent before we can be done with this PR.

I reckon we have a couple of options to consider here:

1. We could expand the `Hanami::Action::Contract` so that it has the same interface as `Params` and `BaseParams`
2. Or I wonder if there's a way we can have a single class provide this "params" interface to the user, and have it be _provided_ the "validation" object (which is either a dry-validation contract or a dry-schema schema)? This way we can stick with a single class for params and not have to maintain two parallel implementations.

I think I like option (2) better. But I'd be keen for your thoughts!

Reply:
Option 2 sounds much better

Reply:
Well going with option 1 we would have easier time to change either of the implementation if there is ever a need for it, change of approach.

But for now it makes no sense to "guess" that it might happen, so I like option 2 better too.

I'll try to implement it in the next couple of days.

Thanks for this I completly missed those few methods I did wanted them to have interface as close as possible.

Reply:
Thanks @krzykamil! Looking forward to seeing how you go, and of course I'm happy to provide any feedback you might need along the way.

Reply:
@timriley I've pushed the changes in the last two commits.

The idea is:

Common functionality/interface can be divided into two modules:
1. Validations handling
2. Base functions

So I created two modules, included in BaseParams and ContractClass (not in Params since it inherits from BaseParams). One of them includes the conversion methods like to_h, and getters like dig, get [] etc. Those are the base functions.
The other module handles the validations parts, extracting params, handling and parsing errors, the more specific stuff connected to Dry::Validation. It also sets the same way to initialize both Params and Contract classes.

This left both Contract and Params classes as rather slim 
Params only has  
self._base_rules
self.params(&blk)

Those two class methods left. Contract respectiely has self.contract but also some extra stuff that Params class gets from the hanami validations gem, that I could not get around. 

Also take note of the _base_rules, I did not implement an adequate method for Contract class, as I wasn't sure if we should have that there and if yes, what is the best way to do it (would appreciate a suggestion)

Anyway, let me know if this approach is OK, what else can be modified and improved.

I've copied a lot of specs from Params specs, to ensure their interfaces are indeed the same and they react similarly to similar operations